File: favouriteGenre.py
def favoriteGenre(userMap, genreMap):
    songtogenre={}

    for genre,songs in genreMap.items(): #O(n)
        for song in songs:
            songtogenre[song]=genre
    
    res={}

    for user,songs in userMap.items(): #O(m) - so in total O(m*n) (n from line 14)
        res[user]=[]
        max_count=0
        countmap={}
        for song in songs: #O(n)
            genre=songtogenre[song]
            countmap[genre]=countmap.get(genre,0)+1
            max_count=max(max_count,countmap[genre])

        for genre,count in countmap.items():
            if count==max_count:
                res[user].append(genre)
    return res


userSongs = {
    "David": ["song1", "song2", "song3", "song4", "song8"],
    "Emma": ["song5", "song6", "song7"]
}

songGenres = {
    "Rock": ["song1", "song3"],
    "Dubstep": ["song7"],
    "Techno": ["song2", "song4"],
    "Pop": ["song5", "song6"],
    "Jazz": ["song8", "song9"]
}

res = favoriteGenre(userSongs, songGenres)
print(res)

File: Favourite Genre.java
# Time Complexity : O(m*n)
# Space Complecity : O(m)
# The code ran on LeetCode
# Maintain a hashmap to get the count of each genres a user listens to

import java.util.*;

public class GenreAnalyzer {

    public static Map<String, List<String>> favoriteGenre(Map<String, List<String>> userMap, Map<String, List<String>> genreMap) {

        HashMap<String, String> songToGenre = new HashMap<>();
        for (String genre : genreMap.keySet()) {
            List<String> songs = genreMap.get(genre);
            for (String song : songs) {
                songToGenre.put(song, genre);
            }
        }

        Map<String, List<String>> result = new HashMap<>();

        for (String user : userMap.keySet()) {
            result.put(user, new ArrayList<>());
            HashMap<String, Integer> countMap = new HashMap<>();
            int max = 0;
            List<String> songs = userMap.get(user);
            for (String song : songs) {
                String genre = songToGenre.get(song);
                countMap.put(genre, countMap.getOrDefault(genre, 0) + 1);
                max = Math.max(max, countMap.get(genre));
            }

            for (String genre : countMap.keySet()) {
                if (countMap.get(genre) == max) {
                    result.get(user).add(genre);
                }
            }
        }
        return result;
    }

    public static void main(String[] args) {
        HashMap<String, List<String>> userSongs = new HashMap<>();
        userSongs.put("David", Arrays.asList(new String[]{"song1", "song2", "song3", "song4", "song8"}));
        userSongs.put("Emma", Arrays.asList(new String[]{"song5", "song6", "song7"}));

        HashMap<String, List<String>> songGenres = new HashMap<>();
        songGenres.put("Rock", Arrays.asList(new String[]{"song1", "song3"}));
        songGenres.put("Dubstep", Arrays.asList(new String[]{"song7"}));
        songGenres.put("Techno", Arrays.asList(new String[]{"song2", "song4"}));
        songGenres.put("Pop", Arrays.asList(new String[]{"song5", "song6"}));
        songGenres.put("Jazz", Arrays.asList(new String[]{"song8", "song9"}));

        Map<String, List<String>> res = favoriteGenre(userSongs, songGenres);

        System.out.println(res);
    }
}
